Nicer looking, more consistent form/instance lists

cooperka opened the pull request:
Closes https://github.com/opendatakit/roadmap/issues/29

Comprehensive before/after screenshots can be seen on Google Drive [here](https://drive.google.com/open?id=1LcxscAB5OlXJKbpdWmJFOwKWHIHw0FIB). One example:

<div>
<img width="180" alt="Delete Saved (before)" src="https://user-images.githubusercontent.com/2047062/52128653-85058680-2603-11e9-93b9-c83465dc7ecc.png">

<img width="180" alt="Delete Saved (after)" src="https://user-images.githubusercontent.com/2047062/52128652-85058680-2603-11e9-8b5b-3d9b953bd68f.png">
</div>

*Delete Saved before/after*

Note: The screenshots on Drive have a blue icon for "finalized" forms but the latest icon is deep purple as shown in the screenshot above.

Several layout files were also renamed and/or consolidated; see the end for a table describing the changes.

#### Resources

- [Material Design 'list' spec](https://material.io/design/components/lists.html#specs)
- [Material Design 'list item divider' spec](https://material.io/design/components/dividers.html#specs)
- [Material Design 'icons' spec](https://material.io/design/iconography/system-icons.html#design-principles)
- [Material Design color palette](https://material-ui.com/style/color/#color-palette)

#### What has been done to verify that this works as intended?

Manual testing on every screen of the app:

- Fill Blank Form
- Edit Saved Form
- Send Finalized Form
- View Sent Form
- Get Blank Form
- Delete Saved Form: Saved Forms
- Delete Saved Form: Blank Forms

With every possible form state:

- Blank, not yet an instance
- Blank with updates available
- Saved
- Finalized
- SMS submission in progress
- Submitted
- Submission Failed

In every possible mode:

- SMS submissions
- Google Drive forms
- Toggle "Show Sent and Unsent Forms" on Send Finalized
- Light/dark themes
- Portrait/landscape

And down to API level 16.

If I missed anything, please let me know and I'll add to the list.

#### Why is this the best possible solution? Were any other approaches considered?

This updates every form list in the app to conform with Material Design specs without any change in behavior. It also consolidates several redundant layouts and classes.

The new icons use Material Design's blue/green/red/deepPurple/amber **500** for icon colors, which is consistent with the existing blue icons and status bar color already used throughout the app.

See the discussion on the [roadmap](https://github.com/opendatakit/roadmap/issues/29) for more context on why these particular icons were decided on.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The behavior of the app shouldn't be affected by this PR, aside from disabled form items showing a Toast showing the reason they're disabled (e.g. if they've been encrypted or deleted from the device). Previously there was no Toast for these items.

#### Do we need any specific form for testing your changes? If so, please attach one.

No specific form, just various form states.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

Probably updating screenshots? TODO

#### Table of changed layouts

Several layout files were renamed and/or consolidated:

| Screen | Main class | Screen layout (OLD) | Screen layout (NEW) | List item layout (OLD) | List item layout (NEW) |
|---|---|---|---|---|---|
| Fill Blank | FormChooserList | chooser_list_layout | form_chooser_list | two_item | form_chooser_list_item |
| View Saved | InstanceChooserList | chooser_list_layout | form_chooser_list | two_item | form_chooser_list_item |
| Send Finalized | InstanceUploaderList | instance_uploader_list | instance_uploader_list | instance_upload_list_item | form_chooser_list_item_multiple_choice |
| View Sent | InstanceChooserList | chooser_list_layout | form_chooser_list | two_item | form_chooser_list_item |
| Get Blank (default) | FormDownloadList | remote_file_manage_list | form_download_list | two_item_multiple_choice | form_chooser_list_item_multiple_choice |
| Get Blank (with Drive) | GoogleDriveActivity | drive_layout | google_drive_list | two_item_image | form_chooser_list_item_multiple_choice |
| Delete Saved (abstract) | FileManagerTabs | file_manager_layout | file_manager_tabs | - | - |
| Delete Saved: Saved | FileManagerFragment (DataManagerList) | tab_layout | file_manager_list | two_item_multiple_choice | form_chooser_list_item_multiple_choice |
| Delete Saved: Blank | FileManagerFragment (FormManagerList) | tab_layout | file_manager_list | two_item_multiple_choice | form_chooser_list_item_multiple_choice |

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

The new icons are entirely composed of [Material Design (MD) icons](https://material.io/tools/icons/?style=baseline), though they have been combined in some cases (e.g. MD checkmark inside MD cloud inside a circle). MD's [Apache 2.0 license](https://tldrlegal.com/license/apache-license-2.0-(apache-2.0)) permits modification. Whether this particular modification is "significant" enough to warrant additional specific mention on the [attributions page](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html) is up to the reviewers.

cooperka committed:
Rename form lists for consistency

cooperka committed:
Format form lists consistently for better future diffs

View this diff with '-w' flag or 'w=1' URL param.

cooperka committed:
Rename form list items for consistency

cooperka committed:
Format form list items consistently for better future diffs

View this diff with '-w' flag or 'w=1' URL param.

cooperka committed:
Give form lists Material Design 8dp vertical padding

cooperka committed:
Remove form 'visibility' icon; disable row and show toast

cooperka committed:
Set 38% opacity for disabled form items

The text color itself may also be changing in an upcoming commit per
Material guidelines.

cooperka committed:
Tweak theme colors to conform to Material specs

cooperka committed:
Use Material text and padding for form_chooser_list_item

cooperka committed:
Factor out form_chooser_list_item_text into separate file

cooperka committed:
Remove redundant text alignment properties

So list item text doesn't overlap the image

cooperka committed:
Display form_state_blank icon by default

Other states will be handled soon.

cooperka committed:
Use form_chooser_list_item_text for multiple choice items as well

cooperka committed:
Factor out form_chooser_list_item_icon into separate file

cooperka committed:
Also display form_state_blank icon on multiple_choice item

By default; new states coming soon.

cooperka committed:
Use Material text and padding for multiple choice list items

cooperka committed:
Use form_chooser_list_item_multiple_choice for google drive too

cooperka committed:
Add convenience links to Material spec for future maintainers

cooperka committed:
Also set opacity on disabled image view

cooperka committed:
Set item icon based on form status for View Sent

cooperka committed:
Use InstanceListCursorAdapter for both 'sent' and 'saved'

cooperka committed:
Add inset list_item_divider to every form list

cooperka committed:
Factor out form_chooser_list_item_checkbox into separate file

cooperka committed:
Use form_chooser_list_item_text for Send Finalized

cooperka committed:
Remove Send Finalized CardView wrapper

cooperka committed:
Remove Send Finalized dimen abstractions for clarity

They'll be changing in upcoming commits, easier to see them here.

cooperka committed:
Use form_chooser_list_item_icon for Send Finalized

cooperka committed:
Use Material Design layout for Send Finalized

`instance_uploader_list_item` now mirrors `form_chooser_list_item_multiple_choice`,
just with a few extra components.

cooperka committed:
Simplify form lists using LinearLayout

cooperka committed:
Use generic form_chooser_list_item_multiple_choice for Send Finalized

cooperka committed:
Use the new icons for Send Finalized

cooperka committed:
Use the new icons for Delete Saved Instances

cooperka committed:
Make TwoItemMultipleChoiceView a LinearLayout

Gets rid of one extra layer of nesting

cooperka committed:
Use Material Design 'cancel' icon for sending SMS

cooperka committed:
Use srcCompat for lower API compatibility

cooperka committed:
Use light/dark themed list item divider colors

cooperka committed:
Shrink the 'cancel' icon down to a proper size

cooperka committed:
Remove now-unnecessary list_item_checkbox layout

cooperka committed:
Make 'finalized' icon indigo to differentiate it

yanokwa commented:
@zestyping I can't seem to add you as a reviewer, but yeah, we'd love a review of this!

zestyping commented on collect_app/src/main/res/layout/form_chooser_list_item_text.xml:
```diff
@@ -0,0 +1,56 @@
+    <!-- Material Design reference: https://material.io/design/components/lists.html#specs -->
+
+    <TextView
+            android:id="@+id/text1"
```
Why do these TextViews have numbered IDs `text1`, `text2`, `text3`, `text4`, and why are they out of order (1, 3, 2, 4)?  Is it possible to give them descriptive IDs according to their intended purpose?

cooperka replied:
These IDs have apparently been like this for [over 7 years](https://github.com/opendatakit/collect/blame/78281ed4389a96425c88f7bbe20df71e732a2f94/collect_app/src/main/res/layout/two_item.xml) -- I have no idea why, but I agree it's confusing.

They're used for different things in different views which is why I didn't touch them, but I will give them generic names at least. Perhaps `title`, `subtitleA`, `subtitleB`, etc.

The main point of confusion is that `text3` is **above** `text2` because it's an optional extra item that most views don't have, but the ones that do have it want it on top.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/adapters/InstanceListCursorAdapter.java:
```diff
@@ -83,9 +98,29 @@ public View getView(int position, View convertView, ViewGroup parent) {
+                imageView.setImageResource(R.drawable.form_state_submitted);
+                break;
+            case InstanceProviderAPI.STATUS_SUBMISSION_FAILED:
+                imageView.setImageResource(R.drawable.form_state_failed);
```
`form_state_submitted` and `form_state_failed` are the cloud icons (green and red respectively), right?  `failed` by itself could mean several things; how about more specific names `form_state_uploaded` and `form_state_upload_failed`?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/adapters/InstanceListCursorAdapter.java:
```diff
@@ -63,15 +73,20 @@ public View getView(int position, View convertView, ViewGroup parent) {
-        visibleOff.setScaleX(0.9f);
-        visibleOff.setScaleY(0.9f);
         if (date != 0 || !formExists || isFormEncrypted) {
+            view.setEnabled(false);
```
Maybe this if/else statement would be a good bit to factor out into a small `setEnabled(View view, boolean enabled)` method?  It looks like `titleText` and `subtitleText` and `visibilityOffCause` are only used in this section, so they could be local variables in the method.

zestyping reviewed (commented):
Thanks for doing all this cleanup!  The new names are so much better, the layout looks better, and the form state icons are a great improvement.

I've made a few small suggestions in the code; nothing major.

I mostly like the icons, but I have concerns with two of them:

* The "finalized" icon and the "uploaded" icon both use the same check mark, and I fear that may be confusing.  If the check mark means "something was attempted, and it succeeded" then it pairs with the exclamation mark, and only makes sense for the "uploaded" icon.  It seems the chief meaning of the "finalized" icon is that you can't edit the form anymore, which could be conveyed nicely by a small padlock; could there be a small lock icon in place of the checkmark?

* The "sending" icon is a chat bubble, which doesn't look like any of the other form icons.  Since it is related to network transmission, it seems like it should resemble the other cloud icons.  How about a cloud with a simplified clock face inside, to indicate that transmission is scheduled?

cooperka commented:
Thank you for the feedback @zestyping 🍻 I'm going to move your feedback about the icons to the proposal if you don't mind, so that all the icon discussions are in one place. I'll tag you and reply there.

cooperka committed:
Rename form_state_{,submission_}failed for clarity

cooperka committed:
Factor out setEnabled/Disabled methods

cooperka committed:
Make views 'final' to appease checkstyle

cooperka committed:
Remove useless gravity specifications for text

It was only on some of them anyway.

cooperka committed:
Swap text2, text3 so they're in the right order

Only 2 screens use text3, and both use VersionHidingCursorAdapter.
They're the only ones that use VHCA.

The only additional change needed is to make sure text3 is made visible
in VHCA because text2 always used to be visible and now they're swapped.

cooperka committed:
Rename text1 => form_title

cooperka committed:
Rename text2 => form_subtitle

cooperka committed:
Remove redundant text4 view

cooperka committed:
Rename text3 => form_subtitle2

cooperka committed:
Rename update_info => form_update_alert

cooperka commented:
Updated old order and naming conventions across the app.

| Old name | New name       |
|----------|----------------|
| text1    | form_title     |
| text2    | form_subtitle  |
| text3    | form_subtitle2 |
| text4    | *[removed]*    |

zestyping commented:
Thanks for the fixes!

cooperka commented:
@zestyping do you think this is ready for testing, or are you planning to give it another look first?

zestyping commented:
@cooperka It looks good to me, go ahead!

zestyping commented:
Oh, sorry, maybe I'm not clear on the review process.  I thought it was up to you (the author) to tag the PR "needs testing".  Is the author or the reviewer supposed to be the one that sets the "needs testing" tag?

cooperka commented:
Awesome, thanks. It looks like it's the reviewer who should label (based on [CONTRIBUTING.md](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#the-review-process)). Then the tester will change the label based on the test results.

I got it this time though 😉 @opendatakit-bot label "needs testing"

grzesiek2010 commented:
@cooperka you as an author shouldn't add `needs testing` labels. @zestyping you are a reviewer so you should accept changes at least before. I very rarely add `needs testing` label to my own prs and it happens only if it's something small an obvious like: https://github.com/opendatakit/collect/pull/2840

grzesiek2010 commented:
@cooperka 

| Blank form  | Saved form | Finalized form |
| ------------- | ------------- | ------------- |
| ![blank](https://user-images.githubusercontent.com/3276264/52061831-8be2b980-256f-11e9-942d-86c8231a7601.png)  | ![saved](https://user-images.githubusercontent.com/3276264/52061833-8be2b980-256f-11e9-8243-57fbc06b27df.png)  | ![finalized](https://user-images.githubusercontent.com/3276264/52061832-8be2b980-256f-11e9-8f68-458377712d91.png)  |

I think it would be better if all those three icons have the same color. What do you think?

grzesiek2010 commented:
The bar at the bottom seems redundant now. The icon and text indicates it's been sent. Do you agree?
![screenshot from 2019-01-31 15-53-57](https://user-images.githubusercontent.com/3276264/52062323-9cdffa80-2570-11e9-97de-6ef19c8f171e.png)

grzesiek2010 commented:
There is empty spce between the app bar and the list:

| This pr  | Master branch |
| ------------- | ------------- |
| ![screenshot_2019-01-31-16-00-38](https://user-images.githubusercontent.com/3276264/52063113-3b209000-2572-11e9-9415-d93a165eee82.png) | ![screenshot_2019-01-31-16-04-56](https://user-images.githubusercontent.com/3276264/52063114-3bb92680-2572-11e9-9e50-1d40105fd4d8.png)  |

The same at the bottom.

grzesiek2010 commented on collect_app/src/main/res/drawable/form_state_submission_failed.xml:
```diff
@@ -0,0 +1,17 @@
+
+    <path
+            android:pathData="M60,0C26.88,0 0,26.88 0,60c0,33.12 26.88,60 60,60 33.12,0 60,-26.88 60,-60 0,-33.12 -26.88,-60 -60,-60"
+            android:fillColor="#F44336"/>
```
Please use `@color/red` here instead of `#F44336` since exactly the same color is already defined in our colors.xml file.

cooperka replied:
👍 I'll make all these new colors into variables like this one for consistency.

grzesiek2010 commented on collect_app/src/main/res/drawable/form_state_blank.xml:
```diff
@@ -0,0 +1,14 @@
+        android:viewportHeight="120">
+
+    <path
+            android:fillColor="#2196F3"
```
You can use `lightAccentColor` here.

grzesiek2010 commented on collect_app/src/main/res/drawable/form_state_saved.xml:
```diff
@@ -0,0 +1,17 @@
+
+    <path
+            android:pathData="M60,60m-60,0a60,60 0,1 1,120 0a60,60 0,1 1,-120 0"
+            android:fillColor="#2196F3"/>
```
You can use `lightAccentColor` here.

grzesiek2010 commented:
As you can see in the last comment https://github.com/opendatakit/collect/pull/2825#issuecomment-459380216

form names are much smaller. It's understandable since you added icons and have less space but don't you think it's too small in comparison with the subtext? It could be a bit bigger maybe?

cooperka commented:
> I think it would be better if all those three icons [blank, saved, finalized] have the same color

Originally they were all the same, but we realized it's hard to tell the difference between "saved" and "finalized" without a difference in color. The reason we wanted to add icons in the first place was to make it much clearer to the user when forms were finalized.

Note that "blank" forms and "saved" forms never appear in the same list together, so now every form state on every screen has its own unique color.

> The [SMS progress] bar at the bottom seems redundant now

I agree it's redundant, but I didn't want to modify this logic too much and have more to test in one PR. The old UI also had a check mark and progress bar, so I left it. Can you please open a new issue for this?

> There is empty space between the app bar and the list

MD lists should have 8dp padding at the top/bottom, but it's meant to be part of the first list item, and **scroll** with the whole list. I'll try to address this.

> Form names are much smaller

All fonts and font sizes used here are MD compliant. If the fonts are too hard for users to read, they're encouraged to increase their Android system font size.

Thank you for the thorough review 😎 I'll address these shortly.

cooperka committed:
Disable clipToPadding for all lists

So the MD 8dp padding doesn't cover the actual list items.

cooperka committed:
Add MD palette to colors.xml

cooperka commented:
@opendatakit-bot label "needs review"

grzesiek2010 commented:
>Originally they were all the same, but we realized it's hard to tell the difference between "saved" and "finalized" without a difference in color. The reason we wanted to add icons in the first place was to make it much clearer to the user when forms were finalized.

Note that "blank" forms and "saved" forms never appear in the same list together, so now every form state on every screen has its own unique color.

Ok makes sense.

> I agree it's redundant, but I didn't want to modify this logic too much and have more to test in one PR. The old UI also had a check mark and progress bar, so I left it. Can you please open a new issue for this?

I'll do that.

> All fonts and font sizes used here are MD compliant. If the fonts are too hard for users to read, they're encouraged to increase their Android system font size.

Right.

Thanks for answering all questions and fixing those things I asked about.

zestyping commented:
> Originally they were all the same, but we realized it's hard to tell the difference between "saved" and "finalized" without a difference in color. 

Oops, that's a surprise to me—when I reviewed, I assumed that the screenshots in this PR description were accurate.  Are they no longer correct?

cooperka commented:
@zestyping everything is the same except the "finalized" icon is deep purple instead of blue to prevent confusion. I just updated the screenshot in the PR description so you can see.

cooperka commented:
@grzesiek2010 this is still under "needs review", should it be ready for testing?

grzesiek2010 commented:
> this is still under "needs review", should it be ready for testing?

yes, we can add `needs testing` label. I wanted to wait and allow other contributors reviewing (since our QAs are busy anyway) but I think that's enough.

yanokwa commented:
Glad to see the progress on this! And speaking of progress, @grzesiek2010, don't forget to file the issue about the progress bar and checkmark.

mmarciniak90 commented:
@cooperka @yanokwa @grzesiek2010 

I tested a lot of cases on Android 6.0.

I noticed that scroll for all pages does not start from the beginning of list but few millimeters under. 
That cause first form area looks bigger than the others.
It is visible for light and dark theme.

I wanted to download forms from Google Drive and I noticed that I'm not able to select the form during the first click. The form flashes but the checkbox is not selected. `Download Selected` button becomes active. When I click on another checkbox the first checkbox is selected.

![checkboxes](https://user-images.githubusercontent.com/3907554/52573791-2b7b3400-2e1b-11e9-8183-63b700940a71.gif)

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

cooperka committed:
Overlay scrollbar for all lists

So MD padding top/bottom doesn't push the scrollbar down.

yanokwa commented:
@cooperka Just to be precise here, the only remaining issues with this PR are the issues @mmarciniak90 noted. We are very close to getting this merged!

cooperka commented:
Thanks @mmarciniak90, the new padding at the top of each list is expected, but I've fixed the scrollbar so it goes all the way to the top now.

~~I don't see how this would have broken the Google Drive forms list, can you confirm that it works on master? I'm going to try setting up the API locally so I can test too.~~

Verified, definitely broken in this PR. I fixed it now.

`FileArrayAdapter` now uses `TwoItemMultipleChoiceView` which controls its own checkbox state, so `GoogleDriveActivity` no longer needs to control that state itself. I cleaned up the logic so it works again (no other activities were affected in this way, just `GoogleDriveActivity`).

cooperka commented:
@opendatakit-bot label "needs testing"

cooperka committed:
Merge branch 'master' into form-list-polish

cooperka committed:
Clean up GoogleDriveActivity setChecked logic

Checkbox is already checked by FileArrayAdapter.

mmarciniak90 commented:
@cooperka 
I'm not able to open any view on Android 4.4, 4.2, 4.1 because of the crash
[crash.txt](https://github.com/opendatakit/collect/files/2876042/crash.txt)

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

cooperka committed:
Merge branch 'master' into form-list-polish

cooperka committed:
Remove unused divider dimens

Tap target not used on ListViews, just Dialogs.

cooperka committed:
Rename divider => dividerCompat for clarity

cooperka committed:
Default to dividerCompat and upgrade on newer devices

Fixes crash in API < 21.

cooperka commented:
Fixed. I ended up defaulting to the old divider and upgrading only if the device supports it.

More detail: Devices < API 21 were failing with `Resources$NotFoundException: File res/drawable/list_item_divider.xml`. Normally this would be solved by a workaround like using `srcCompat` instead of `src` but in this case Android doesn't provide a `dividerCompat` option. The alternative workarounds are described well in [this StackOverflow diagram](https://stackoverflow.com/a/41855830/763231); we already have `vectorDrawables.useSupportLibrary` and I didn't want to `setCompatVectorFromResourcesEnabled` so this is the best solution I could come up with.

@opendatakit-bot label "needs testing"

mmarciniak90 commented:
**Tested with success**

Tested on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Tested cases: 
* Fill Blank Form: blank status icon
* Edit saved form: saved, finalized, failed statuses icons
* Send Finalized form  / Show Unsent Forms: finalized, failed statuses icons
* Send Finalized form  / Show Sent and Unsent Forms: sent, finalized, failed statuses icons
* View Sent Form: visible, not visible icons
* Get Blank Form / Google Drive
* Get Blank Form / Aggregate
* Delete Saved Form / Saved Forms: saved, finalized, failed statuses icons
* Delete Saved Form / Blank Forms: blank status icon
---
* SMS submission (with/without success, no sms tags, airplane mode)
* Aggregate submission
* Google Drive submission
---
* Version number is visible
* Green information that new version is ready to update
* Selected forms are not mixed
* Horizontal and vertical device orientation
* Light and dark mode
* Scroll starts from the beginning
* Sorting

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
